### PR TITLE
Checkout and build opensim-core for github Action

### PR DESCRIPTION
**Pull request by aymanhab**
Fixes issue #<issue_number>

### CHANGELOG.md (choose one)

- no need to update because...
- updated...

### CODE REVIEW

**Comment by aymanhab** [comment]
@chrisdembia I can checkout and build opensim-core in github action on windows but the build gets stuck, any ideas?

**Comment by aymanhab** [comment]
@halleysfifthinc would you be able to help get GitHub Actions to work on linux? Thanks much

**Comment by halleysfifthinc** [comment]
I will definitely take a quick look, but I won't have much/any time to spend on this till August.

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 name: opensim-application
 
-on: [push]
+on:
+  push:
+    branches: 
+      - 'action_build_core'
```
Comment (halleysfifthinc) [inline .github/workflows/continuous-integration.yml:6]: I believe the previous behavior ran on any push (or possibly only to the master branch); possibly a good idea to also have it triggered for pushes to master?
```suggestion
      - 'action_build_core'
      - 'master'
```

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -9,11 +12,48 @@ jobs:
     runs-on: windows-latest
 
     steps:
-    - name: Install opensim-core
+    - uses: actions/checkout@v1
```
Comment (halleysfifthinc) [inline .github/workflows/continuous-integration.yml:15]: Might as well update this and other occurrences of `actions/checkout@v1`?
```suggestion
    - uses: actions/checkout@v2
```

File: .github/workflows/continuous-integration.yml [hunk 3]
```diff
@@ -121,15 +161,39 @@ jobs:
...
+    - name: Checkout opensim-core master
+      uses: actions/checkout@v2
+      with:
+         name: opensim-org/opensim-core
+         repository: opensim-org/opensim-core
+         path: 'opensim-core'
+    
+    - name: Build opensim-core-dependencies
+      run: |
+        echo $env:GITHUB_WORKSPACE\\build_deps
```
Comment (halleysfifthinc) [inline .github/workflows/continuous-integration.yml:178]: Commands run from the GITHUB_WORKSPACE folder by default; I also don't believe the `env:` is necessary, `$GITHUB_WORKSPACE` should work.
Also, the path separator on Linux is forward-slash `/`.

File: .github/workflows/continuous-integration.yml [hunk 4]
```diff
@@ -121,15 +161,39 @@ jobs:
...
+        mkdir $env:GITHUB_WORKSPACE\\build_deps
+        chdir $env:GITHUB_WORKSPACE\\build_deps
+        $env:CXXFLAGS = "/W0"
```
Comment (halleysfifthinc) [inline .github/workflows/continuous-integration.yml:181]: Environment variables can be set in the Actions yml config; see [Using Environment Variables](https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables)

File: .github/workflows/continuous-integration.yml [hunk 5]
```diff
@@ -121,15 +161,39 @@ jobs:
...
+      with:
+         name: opensim-org/opensim-core
+         repository: opensim-org/opensim-core
+         path: 'opensim-core'
+    
+    - name: Build opensim-core-dependencies
+      run: |
+        echo $env:GITHUB_WORKSPACE\\build_deps
+        mkdir $env:GITHUB_WORKSPACE\\build_deps
+        chdir $env:GITHUB_WORKSPACE\\build_deps
```
Comment (halleysfifthinc) [inline .github/workflows/continuous-integration.yml:180]:
```suggestion
        cd build_deps
```

**Comment by aymanhab** [comment]
Windows work now, merging while keeping branch to fix linux

**Comment by aymanhab** [comment]
@chrisdembia windows now work in github actions (need caching for performance) but linux fails due to simbody-visualizer not found, any clues?

**Comment by marnunez** [comment]
> @chrisdembia windows now work in github actions (need caching for performance) but linux fails due to simbody-visualizer not found, any clues?


I've encountered this before: there is some discrepancy between building the GUI against the latest opensim-core bundle from Sourceforge and building it against opensim-core master. I belive the SDK folder is not created automatically in opensim-core unless specified in CMake. Could it be that? There's something the old linux build does right that the newer doesn't
By the way: solving this would also solve opensim-gui being built against and old Python 3.4 release. Nice.

**Comment by halleysfifthinc** [comment]
opensim-core needs to be built with `-DOPENSIM_INSTALL_UNIX_FHS=OFF` for the GUI. This is currently the case for the deployed builds of opensim-core (based on [.travis.yml](https://github.com/opensim-org/opensim-core/blob/master/.travis.yml)). On *nixes, this defaults to `ON`, so if you build opensim-core yourself, you will need to specify the alternative when you run cmake.

**Comment by marnunez** [comment]
> opensim-core needs to be built with `-DOPENSIM_INSTALL_UNIX_FHS=OFF` for the GUI. This is currently the case for the deployed builds of opensim-core (based on [.travis.yml](https://github.com/opensim-org/opensim-core/blob/a470a0e7748997f87239cad4dd98eee1d2a7bb2a/.travis.yml#L218)). On *nixes, this defaults to `ON`, so if you build opensim-core yourself, you will need to specify the alternative when you run cmake.

:facepalm:

**Comment by aymanhab** [comment]
Thanks for the pointer @marnunez and @halleysfifthinc 👍 

I made the change to see if it goes thru and we get an artifact but long term we should make the CMake defaults work out of the box i.e. make GUI build assume OPENSIM_INSTALL_UNIX_FHS=ON or change the flag to OFF by default if it has no other downstream side-effects.

**Comment by chrisdembia** [comment]
> we should make the CMake defaults work out of the box i.e. make GUI build assume OPENSIM_INSTALL_UNIX_FHS=ON or change the flag to OFF by default if it has no other downstream side-effects.

I think the correct approach is that the GUI's CMakeLists.txt should ensure opensim-core was built with `OPENSIM_INSTALL_UNIX_FHS=OFF`. I think it's useful to have `OPENSIM_INSTALL_UNIX_FHS=ON` as the default in opensim-core, because Linux users would be surprised by an installation layout that does not conform to the Filesystem Hierarchy Standard, especially because users may install opensim-core system-wide.

**Comment by aymanhab** [comment]
Thanks for chiming in @chrisdembia but the GUI does not "build" opensim-core internally, instead it assumes it was already built. Can you explain the rationale for building the GUI against the non-preferred layout on linux, rather than modifying the build script?

**Comment by chrisdembia** [comment]
> Thanks for chiming in @chrisdembia but the GUI does not "build" opensim-core internally, instead it assumes it was already built.

I understand. I meant to say that the GUI can raise an error if it detects that opensim-core was built with `OPENSIM_INSTALL_UNIX_FHS=ON`.

> Can you explain the rationale for building the GUI against the non-preferred layout on linux, rather than modifying the build script?

When you and I discussed the layout of the GUI on Mac, we confronted this same issue. I originally thought it would be better to use the UNIX FHS layout for the GUI on Mac, but I think we mutually arrived at the conclusion that using the same layout of the GUI across platforms was the right way to go.

**Comment by aymanhab** [comment]
Thanks @chrisdembia I opened issue https://github.com/opensim-org/opensim-gui/issues/1200 to continue the discussion of desired behavior.

**Comment by aymanhab** [comment]
With artifacts from windows and linux available here https://github.com/opensim-org/opensim-gui/actions/runs/114987515 this PR is ready for review. We can try adding caching for opensim-core now or do it later when commits on this repo are more frequent. Would be great if somebody on ubuntu can test the artifact as well.

File: .github/workflows/continuous-integration.yml [hunk 6]
```diff
@@ -1,6 +1,10 @@
 name: opensim-application
 
-on: [push]
+on:
+  push:
+    branches: 
+      - 'action_build_core'
+      - 'master'
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:7]: I suggest we use what we do for Moco:

```
on:
  pull_request:
    branches:
      - '*'
  push:
    branches: 
      - master
    tags:
      - '*'
```

File: .github/workflows/continuous-integration.yml [hunk 7]
```diff
@@ -9,11 +13,51 @@ jobs:
...
+
+    - name: Build opensim-core
+      run: |
+        echo $env:GITHUB_WORKSPACE\\build_core
+        mkdir $env:GITHUB_WORKSPACE\\build_core
+        chdir $env:GITHUB_WORKSPACE\\build_core
+        $env:CXXFLAGS = "/W0"
+        cmake $env:GITHUB_WORKSPACE/opensim-core -G"Visual Studio 16 2019" -A x64 -DOPENSIM_DEPENDENCIES_DIR=~/opensim_dependencies_install -DBUILD_JAVA_WRAPPING=on -DBUILD_PYTHON_WRAPPING=on -DOPENSIM_C3D_PARSER=ezc3d -DBUILD_TESTING=off -DCMAKE_INSTALL_PREFIX=~/opensim-core-install
+        cmake . -LAH
+        cmake --build . --config Release -- /maxcpucount:8
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:54]: Any reason for using 8 here and 4 above?

File: .github/workflows/continuous-integration.yml [hunk 8]
```diff
@@ -9,11 +13,51 @@ jobs:
...
+    - name: install opensim-core
       run: |
-        (New-Object System.Net.WebClient).DownloadFile("https://prdownloads.sourceforge.net/myosin/github-actions/opensim-core/opensim-core-win.zip", "opensim-core-win.zip")
-        7z x opensim-core-win.zip -oC:\\
-        dir C:\\opensim-core
+        chdir $env:GITHUB_WORKSPACE\\build_core
+        cmake --install . 
+
     - name: Checkout opensim-gui
       uses: actions/checkout@v1
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:62]: This is redundant with the first step above.

File: .github/workflows/continuous-integration.yml [hunk 9]
```diff
@@ -121,15 +165,38 @@ jobs:
...
 
     steps:
-    - name: Install opensim-core
-      run: |
-        wget https://prdownloads.sourceforge.net/myosin/opensim-core/opensim-core-latest_linux_Release.zip
-        # -q: quiet
-        unzip -q opensim-core-latest_linux_Release.zip -d ~
 
     - name: Checkout opensim-gui
       uses: actions/checkout@v1
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:170]: There are inconsistencies between platforms related to the step for checking out `opensim-gui`.

File: .github/workflows/continuous-integration.yml [hunk 10]
```diff
@@ -121,15 +165,38 @@ jobs:
...
+    - name: Install packages
+      run: sudo apt-get install --yes liblapack-dev freeglut3-dev libxi-dev libxmu-dev doxygen python3 python3-dev python3-numpy python3-setuptools swig
+
+    - name: Build opensim-core-dependencies
+      run: |
+        mkdir build_deps
+        cd build_deps
+        cmake ../opensim-core/dependencies -DSUPERBUILD_ezc3d:BOOL=on -DCMAKE_INSTALL_PREFIX=~/opensim_dependencies_install
+        cmake . -LAH
+        cmake --build . --config Release
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:189]: I'm not sure `--config` has meaning for generators other than MSVC and Xcode. It's more typical to put `-DCMAKE_BUILD_TYPE=Release` when configuring.

File: .github/workflows/continuous-integration.yml (hunk 7 above)
Comment (halleysfifthinc) [reply .github/workflows/continuous-integration.yml:54]: FYI The Actions CI runners only have 2 cores.

Comment (aymanhab) [reply .github/workflows/continuous-integration.yml]: Thanks @chrisdembia and @halleysfifthinc changed to 2

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (aymanhab) [reply .github/workflows/continuous-integration.yml]: Done

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (aymanhab) [reply .github/workflows/continuous-integration.yml]: Now all use checkout@v2

File: .github/workflows/continuous-integration.yml (hunk 10 above)
Comment (aymanhab) [reply .github/workflows/continuous-integration.yml:192]: Left alone since it works as is and these experiments are expensive.

**Comment by aymanhab** [comment]
Thanks for the feedback @chrisdembia I addressed it, ready for re-review.

File: .github/workflows/continuous-integration.yml [hunk 11]
```diff
@@ -9,13 +17,53 @@ jobs:
...
       run: |
-        (New-Object System.Net.WebClient).DownloadFile("https://prdownloads.sourceforge.net/myosin/github-actions/opensim-core/opensim-core-win.zip", "opensim-core-win.zip")
-        7z x opensim-core-win.zip -oC:\\
-        dir C:\\opensim-core
+        chdir $env:GITHUB_WORKSPACE\\build_core
+        cmake --install . 
+
     - name: Checkout opensim-gui
-      uses: actions/checkout@v1
+      uses: actions/checkout@v2
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:66]: The GUI was already checked out in the first step; no need to check out again.

File: .github/workflows/continuous-integration.yml [hunk 12]
```diff
@@ -75,7 +123,7 @@ jobs:
         # -q: quiet
         unzip -q opensim-core-mac.zip -d ~
     - name: Checkout opensim-gui
-      uses: actions/checkout@v1
+      uses: actions/checkout@v2
```
Comment (chrisdembia) [inline .github/workflows/continuous-integration.yml:126]: Can the location of the step that checks out the GUI be consistent between all platforms?

**Comment by aymanhab** [comment]
I addressed your latest feedback, once windows and linux build successful will merge. I didn't work on OSX, will do later following the linux workflow mostly. Thanks for the review and feedback @chrisdembia

**Comment by chrisdembia** [comment]
LGTM.

We should fix the Mac build as soon as we can, because the Travis Mac build is also failing.

**Comment by aymanhab** [comment]
Thanks much @chrisdembia Merging

**Pull request merged**